sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,571 @@
+kernels can be used to access program cache. However when shared library
+defines kernels and these kernels are run by the application unchanged cache
+structure may lead to double compilation of the same code. Example of code
+that leads to double compilation of library code:
+```
+// Application
+SYCL_EXTERNAL void LibFunc();
+
+Q.submit([&](cl::sycl::handler &CGH) {
+  handler.parallel_for([] { LibFunc(); });  // Device code for library is compiled
+                                            // and linked together with device
+                                            // code for application, i.e.
+                                            // LibFunc1 and ExternalKernel exist
+                                            // in prepared state
+});
+// ...
+EnqueueLibraryKernel(Q); // If cache mechanism is not changed, this line will
+                         // lead to second compilation of ExternalKernel and
+                         // LibFunc
+
+// Library
+SYCL_EXTERNAL void LibFunc() {
+// ...
+}
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+```
+Such case can be optimized by bringing nesting into cache keys structure.
+Kernel set id can be found for each kernel using its name and OS module it is
+coming from. In presence of dynamic linking resulting program can be combined
+out of device images which come from different OS modules. So, it should be
+possible to find needed program by kernel name and any OS module that was
+involved in this program. The new mapping structure is:
+```
+{kernel name} =>
+  {OSModuleHandle, spec const, opts, dev} => program
+```
+I.e. each kernel name is mapped to a set of tuples that consists of OS module,
+spec constant values, JIT compiler options and device. Then concrete tuple is
+mapped to a program object. Several tuples can be mapped to a same program
+object, they are created during process of compilation and symbols resolution
+for concrete device image. When some program is made through linking of several
+programs created from device images that come from different OS modules,
+for each OS module in cache will be created a tuple with corresponding OS module
+id.
+Example of modified cache structure when dynamic linking is involved:
+```
+// Application
+// OSModule = 1
+
+SYCL_EXTERNAL void LibFunc();
+queue.submit(parallel_for<InternalKernel>( ... ));
+
+Q.submit([&](cl::sycl::handler &CGH) {
+  CGH.parallel_for([] { LibFunc(); });
+});
+
+EnqueueLibraryKernel(q);
+
+// Library
+// OSModule = 2
+
+SYCL_EXTERNAL LibFunc();
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+
+Program cache will have the following structure:
+ "InternalKernel" =>
+   {1, ...} => program 1
+ "ExternalKernel" =>
+   {1, ...} => program 1
+   {2, ...} => program 1
+```
+However the library code will be compiled twice if kernel from the library
+was enqueued before kernels from the application, i.e. in such case:
```

Comment:
If you wanted to address this weakness (about compiling twice), I think you could do so by adding a "program state" to cache key.  To see how this would work, first consider a simple case where there are no shared libraries.  Here are the PI operations that are required:

```
pi_program1 = piProgramCreate(/*spir-v from OSMod 1*/);
/* possibly set values of spec constants */
piProgramCompile(pi_program1, dev, opts);
pi_program2 = piProgramLink(dev, opts, {pi_program1});
```

Note that we're calling `piProgramCompile()` followed by `piProgramLink()`, rather than just calling `piProgramBuild()`.  The reason for doing this will become clear later.

Since we are adding a "program state" to the cache key, we can cache both `pi_program1` and `pi_program2`:  The result of `piProgramCompile()` is an object state program image, while the result of `piProgramLink()` is an executable state program image.  Here are the entries in the cache:

```
[OSMod 1, spec consts, opts, dev, object] => pi_program1
[OSMod 1, spec consts, opts, dev, executable] => pi_program2
```

I think this is the same as what you propose, except I'm adding the "program state" to the cache key and I've cached `pi_program1` in addition to `pi_program2`.

Any future request to run a kernel from OS module 1 on the same device with the same spec constants and the same build options will search the cache for `[OSMod 1, spec consts, dev, executable]`, and it will find the `pi_program2` that we already built.

Now let's consider the case where there is device code in a shared library that needs to be online linked.  Using your example above, here are the PI operations that need to happen:

```
pi_program1 = piProgramCreate(/*spir-v from OSMod 1*/);
/* possibly set values of spec constants */
piProgramCompile(pi_program1, dev, opts);

pi_program2 = piProgramCreate(/*spir-v from OSMod 2*/);
/* possibly set values of spec constants */
piProgramCompile(pi_program2, dev, opts);

pi_program3 = piProgramLink(dev, opts, {pi_program1, pi_program2});
```

And we can put the following items in the cache:

```
[OSMod 1, spec consts, opts, dev, object] => pi_program1
[OSMod 2, spec consts, opts, dev, object] => pi_program2
[OSMod 1, spec consts, opts, dev, executable] => pi_program3
[OSMod 2, spec consts, opts, dev, executable] => pi_program3
```

Again, I think this is the same as what you propose except I have the new "program state" element in the key, and I'm caching some more `pi_program` objects.

If there is a subsequent attempt to submit `ExternalKernel`, the program manager will see that `ExternalKernel` comes from `OSMod 2`, and it will find `pi_program3` in the cache.  Again, this is exactly the same as your proposal.

Now, let's consider the case when `ExternalKernel` is submitted first.  The PI operations are:

```
pi_program1 = piProgramCreate(/*spir-v from OSMod 2*/);
/* possibly set values of spec constants */
piProgramCompile(pi_program1, dev, opts);
pi_program2 = piProgramLink(dev, opts, {pi_program1});
```

And the cache will be:

```
[OSMod 2, spec consts, opts, dev, object] => pi_program1
[OSMod 2, spec consts, opts, dev, executable] => pi_program2
```

Now, the application submits `InternalKernel`, and the PI operations are:

```
pi_program3 = piProgramCreate(/*spir-v from OSMod 1*/);
/* possibly set values of spec constants */
piProgramCompile(pi_program3, dev, opts);

pi_program1 = /* found in cache: [OSMod 2, spec consts, opts, dev, object]*/;

pi_program4 = piProgramLink(dev, opts, {pi_program3, pi_program1});
```

Notice that the redundant compilation of "OSMod 2" is eliminated because the result was in the cache.

Reply:
> Do piProgramCompile and piProgramLink produce a distinct program object instead of modifying the source one?

`piProgramLink` creates a new program object, but `piProgramCompile` does not.  The logic I propose above only depends on `piProgramLink` creating a new program object.

> As for performing build in two steps instead of one (I am not really sure about it), the piProgramBuild could be a bit more optimized in time domain than piProgramCompile plus piProgramLink due to internals of backend.

Yes, this is a good point.  Looking at the implementation of `piProgramLink` in the Level Zero PI plugin, I think that code should check for the case when there is only one input.  We can optimize that case because it is not necessary to call `zeModuleDynamicLink`.

For OpenCL, we need to ask the team that implements the OpenCL driver whether it is much more efficient to call `clBuildProgram` vs. calling both `clCompileProgram` and `clLinkProgram`.

Reply:
Optimization of `piProgramCompile` & `piProgramLink` seems a bit out of the scope of this PR. I might include an improvement with object state programs in the cache though. Should I?

Reply:
I do not have a strong opinion on expanding the cache to include object state as part of this design vs. doing it as a separate task.  If you decide not to do it as part of this design, I think you should enter a Jira to make sure it does happen at some point, though.  If you do this, the Jira could contain a link to this conversation (or copy the conversation into the Jira).

I think you should **not** split the calls to `piProgramBuild()` into separate `piProgramCompile()` / `piProgramLink()` calls unless you check that this won't introduce a performance regression.  The fix I mention above to the Level Zero implementation of `piProgramLink()` is probably a very small 3-liner.  You should also ask the OpenCL team about the efficiency of calling `clBuildProgram` vs. calling both `clCompileProgram` and `clLinkProgram`, though.